### PR TITLE
SDI-24 Offender booking entity builder

### PR DESCRIPTION
**Pull request by andymarke**
Next steps 
1) Load PERSON (and contacts)
2) Change Repository tests to use builder
3) Add more tests around person and contacts

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/Offender.kt [hunk 1]
```diff
@@ -30,7 +30,7 @@ data class Offender(
   @GeneratedValue(generator = "OFFENDER_ID")
   @Id
   @Column(name = "OFFENDER_ID", nullable = false)
-  val id: Long,
+  var id: Long = 0,
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/Offender.kt:34]: assigned after creation

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderVisitBalance.kt [hunk 2]
```diff
@@ -13,6 +16,15 @@ data class OffenderVisitBalance(
...
 
+  @OneToOne
+  @JoinColumn(
+    name = "OFFENDER_BOOK_ID",
+    referencedColumnName = "OFFENDER_BOOK_ID",
+    insertable = false,
+    updatable = false
+  )
+  val offenderBooking: OffenderBooking? = null,
+
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderVisitBalance.kt:27]: Feels like I should be able to combine this with the `@Id` field, but requires another PR to work that out

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderVisitBalance.kt [hunk 3]
```diff
@@ -21,4 +33,19 @@ data class OffenderVisitBalance(
 
   @Column(name = "VISIT_ALLOWANCE_INDICATOR")
   val visitAllowanceIndicator: Boolean? = false,
-) : AuditableEntity()
+) {
+  override fun equals(other: Any?): Boolean {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderVisitBalance.kt:37]: need else offenderBooking would be part of equals etc

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/VisitResourceIntTest.kt [hunk 4]
```diff
@@ -44,9 +49,32 @@ class VisitResourceIntTest : IntegrationTestBase() {
...
   @DisplayName("Create")
   @Nested
   inner class CreateVisitRequest {
+    lateinit var offenderNo: String
+    lateinit var offender: Offender
+    private var offenderBookingId: Long = 0L
+
+    @BeforeEach
+    internal fun createPrisoner() {
+      offender = repository.save(
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/VisitResourceIntTest.kt:65]: same offender is created for each test, we could do this per suite if it was too slow but obvious tests might interfere with each other (which in this case they won't)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderVisitBalance.kt (hunk 2 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderVisitBalance.kt:27]: Im surpirsied it doesnt complain that the same col is mapped twice

**Review by steverendell — APPROVED** [review]
lgtm

**Pull request merged**